### DISCUSSION TITLE
[Naive Receiver] How do I sign data in Foundry?

### DISCUSSION THREAD

**Discussion by 0xalexhex** [tinchoabbate/damn-vulnerable-defi#64] [Help / Q&A / Troubleshooting] [answered]
I think I found how to exploit this contract, but I don't understand how to generate a **NaiveReceiverTest::signature** to pass in **BasicForwarder::execute** function. 

I found the similar vulnerability that been exploited using hardhat and ether.js library(
Image (https://github.com/user-attachments/assets/60c05a63-ed9c-4d64-9b20-729b5ca25b54) )
Exact line to generate signature : 
``` javascript 
const signature = await wallet.signTypedData(domain, types, data);
```
But I don't understand how to get this signature function in Foundry, as there is no existing library that I can import to generate it. These are the last lines of my test that I am stuck on in Foundry:
``` solidity 
    bytes32 memory domain = forwarder.domainSperator();
    bytes32 memory types = forwarder.getRequestTypehash();
    bytes32 memory dataHash = forwarder.getDataHash(request);

    bytes memory signature = ??? ;


    forwarder.execute(request,signature);
```

**Comment by Schereo** [comment, marked as answer]
You can get the signature from the forge cheat code `vm.sign()` as explained here: https://book.getfoundry.sh/cheatcodes/sign.

So basically you would do the following to obtain the signature:

```
(uint8 v, bytes32 r, bytes32 s) = vm.sign(playerPk, eip712Hash);
bytes memory signature = abi.encodePacked(r, s, v);
```

In addition make sure to supply the hash in a EIP721 compliant format. That means to use your `dataHash` and hash it again with the following parameters:

```
bytes32 eip712Hash = keccak256(
            abi.encodePacked("\x19\x01", forwarder.domainSeparator(), dataHash)
        );
```

For more information have a look at the EIP712: https://eips.ethereum.org/EIPS/eip-712
